### PR TITLE
CSV import feature

### PR DESCRIPTION
**Pull request by seatedscribe**
## Description

Now it is possible to import databases from straight CSV text file format and from other password managers (as long as they provide a CSV export feature).

## Motivation and Context

This is a pull request originally [made for KeepassX](https://github.com/keepassx/keepassx/pull/148).

## How Has This Been Tested?

When we talk about security TDD is a must; so I tried hard to put all the conceivable test patterns together in the unit test file, and get to successfully pass them eventually. Please drop me a line if you find some coverage gap! Pattern test of the CSV parser are available in tests/ directory - see TestCsvParser.{h,cpp}

### **PARSER**
I developed the parser with a BNF approach for syntax validation; just for reference, this is the RFC grammar as provided by the IETF (RFC 4180):

file = [header CRLF] record *(CRLF record) [EOF] #[CRLF]
header = name *(COMMA name)
record = field *(COMMA field)
name = field
field = (escaped | non-escaped)
escaped = DQUOTE *(TEXTDATA | COMMA | CR | LF | 2DQUOTE) DQUOTE
non-escaped = *TEXTDATA
COMMA = %x2C
CR = %x0D
DQUOTE = %x22
LF = %x0A
CRLF = CR LF
TEXTDATA = %x20-21 | %x23-2B | %x2D-7E

I needed to validate the two most used versions - the "doubling quotes" and
the "escape character" - into the same engine, so I slightly modified
the grammar to better reflect actual code implementation. I also extended it
to validate the case of "escaped characters". This is the result:

file = record *(CRLF record) EOF
record = field *(separator field) | comment --e.g. ,,A,B has 2 empty fields and is valid
field = simple | quoted
simple = TEXT -- e.g. f""ie"l"""d is OK
quoted = qualifier escaped qualifier
escaped = ESC_TEXT *(escape_mark ESC_TEXT)
escape_mark = (2qualifier | BACKSLASH qualifier)
qualifier = one of {DQUOTE, COLON, BACKSLASH}
separator = one of {COMMA, COLON, SEMICOLON}
comment = *(SPACE | TAB) COMMENT
COMMA = %x2C --i.e. ','
COLON = %x58 --i.e. ':'
SEMICOLON = %x59 --i.e. ';'
COMMENT = %x23 --i.e. '#'
DQUOTE = %x22 --i.e. '"'
BACKSLASH = %x5C --i.e. ''
CR = %x0D
LF = %x0A
SPACE = %x20
CRLF = CR LF | LF
TEXT = *(any char except CRLF, separator, EOF) (also TEXT="" shall be valid)
ESC_TEXT = +(any char except qualifier, EOF)

### **GUI**

I enjoyed putting some effort in GUI design to make the user experience enjoyable and at the same time to be **future-proof** and avoid reinventing the wheel; for further details about this please look at the comment around `CsvImportWidget::m_columnheader`.

- :negative_squared_cross_mark: Bug fix (non-breaking change which fixes an issue)
- :white_check_mark: New feature (non-breaking change which adds functionality)
- :negative_squared_cross_mark: Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- :white_check_mark: I have read the **CONTRIBUTING** document.
- :white_check_mark: My code follows the code style of this project.
- :white_check_mark: All new and existing tests passed.
- :white_check_mark: My change requires a change to the documentation.
- :negative_squared_cross_mark: I have updated the documentation accordingly.
- :white_check_mark: I have added tests to cover my changes.

Best regards,
Enrico

### CODE REVIEW

**Comment by droidmonkey** [comment]
Cursory review of the code looks really good. Did you write the CSV parser yourself?

**Comment by seatedscribe** [comment]
Weird question of yours. Of course I wrote it myself, every single line. Otherwise I would have made it clear who is the author :smiley: 
I really appreciate you find it a good code, I put a lot of design efforts on the GUI and I really enjoyed it!
Looking forward to seeing it merged at last (overall, almost a year has passed since pull request on the original project...)

Regards,
Enrico

**Comment by droidmonkey** [comment]
Yes, just wanted to make sure it was your code. Thank you for the contribution, I will test and approve within a couple days.

File: src/gui/csvImport/CsvImportWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    writer.writeDatabase(&buffer, m_db);
+    if (writer.hasError()) {
+        MessageBox::warning(this, tr("Error"), tr("CSV import: writer has errors:\n").append((writer.errorString())), QMessageBox::Ok, QMessageBox::Ok);
+    }
+    Q_EMIT editFinished(true);
+}
+
+
+Group* CsvImportWidget::grp(QString label)  {
+    Group* root = m_db->rootGroup(), *current = root;
```
Comment (droidmonkey) [inline src/gui/csvImport/CsvImportWidget.cpp]: This does not conform to the declaration style, please put on two separate lines.

File: src/gui/csvImport/CsvImportWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    KeePass2Writer writer;
+    writer.writeDatabase(&buffer, m_db);
+    if (writer.hasError()) {
+        MessageBox::warning(this, tr("Error"), tr("CSV import: writer has errors:\n").append((writer.errorString())), QMessageBox::Ok, QMessageBox::Ok);
+    }
+    Q_EMIT editFinished(true);
+}
+
+
+Group* CsvImportWidget::grp(QString label)  {
```
Comment (droidmonkey) [inline src/gui/csvImport/CsvImportWidget.cpp]: This function name is non-descriptive. Recommend changing to a verb-noun format

File: src/main.cpp [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #include "crypto/Crypto.h"
 #include "gui/Application.h"
 #include "gui/MainWindow.h"
+#include "gui/csvImport/CsvImportWizard.h"
```
Comment (droidmonkey) [inline src/main.cpp]: You have to put this here because you forward-declared the CsvImporter class. Recommend NOT doing that in DatabaseWidget.cpp

**Comment by droidmonkey** [comment]
I built and ran it, everything looks great until I tried to import a CSV file that I just exported from KeePassXC:

csv_import_error

As you can see it totally messed up the import somewhere. I did not run through the debugger to determine what happened but there is something fundamentally broken it would seem.

File: src/gui/MainWindow.ui [hunk 4]
```diff
@@ -126,8 +133,7 @@
     <addaction name="actionChangeMasterKey"/>
     <addaction name="actionChangeDatabaseSettings"/>
     <addaction name="separator"/>
-    <addaction name="actionDatabaseMerge"/>
```
Comment (droidmonkey) [inline src/gui/MainWindow.ui]: You killed the merge database action.

**Comment by TheZ3ro** [comment]
@droidmonkey I think you have to manually set the "Field are separated by" option to "," instead of ";"
And also the Colum Layout of Title, Username, Password, URL, and Notes set to Column2, Column3, etc...

**Comment by droidmonkey** [comment]
Good catch @thez3ro. However the field separator should default to ',' since that is the default in almost all csv exporters.

**Comment by TheZ3ro** [comment]
I agree with @droidmonkey on the default separator :wink:

**Comment by Manko10** [comment]
What about automatically detecting what separator is used? Could be easily done by comparing character frequencies.

**Comment by seatedscribe** [comment]
Thanks a lot for your efforts!

I already made most of the corrections, and I'd like your comment on this:

 @droidmonkey 
- grp method is a short name on purpose: since I am also trying to keep the code aesthetically beautiful to see, a longer name would bring a gaping hole in the last rows of writeDatabase() :smile: Do you think we can accept this - or I need to rename it, like getGroup()?

@TheZ3ro:
- Root base group is actually an unfortunate name for "the group of all the CSV imported items". It should have been named differently, but slipped through.
If you right click on it and "edit group", you'll see a memo of where the data is coming from (and I should also add the time of the operation as well). Its meaning is to keep the user aware that at the end of the day this data is coming from another place/program and not created right into the keepassxc program. In my mind this would raise e.g. that I need to set up the expiration date (which currently is not part of the import process) or things like that.

So we can do two things:

   - change the name of the base group (e.g. CSV IMPORT)
   - skip this "feature" altogether

Let me know your thoughts.

**Comment by TheZ3ro** [comment]
@seatedscribe it's trivial to add an option in the GUI to set the "CSV Root" group name or merge directly in the "Root" tree?
Like a LineEdit for selecting the new group name (that by default is "CSV Import") and a checkbox for using the database "Root" group

What do you think?

**Comment by seatedscribe** [comment]
@TheZ3ro I would rather keep the wizard focused on the CSV data itself, without asking users to make "data presentation" decision at this stage.
I will solve the "double root" issue nevertheless; however, one can easily change the name with a right click once the fields are successfully imported.

**Comment by droidmonkey** [comment]
Re: grp function

It needs to be renamed because it is meaningless and actually hidden in the very long lines of code in WriteDatabase. Lining up code os generally frowned upon because it depends on your font choice in editor anyway. Also, most of thos function calls should probably happen once and cached in a variable outside of the setter calls.

**Comment by TheZ3ro** [comment]
Ping @seatedscribe

**Comment by seatedscribe** [comment]
I am here :smile: 
Just devising a smart way to handle the group root name for any conceivable case. More on this later!

**Comment by seatedscribe** [comment]
Hello again! This update should solve all the issues, please tell me if something else is missing.
I added a check before writing the database in order to create the right root label depending on what is found looking through the group column.

**Comment by droidmonkey** [comment]
PR Closed at request of original submitter. Intent is to resubmit using a different branch.

**Pull request closed without merging**